### PR TITLE
[WIP/RFC] Comprehensive overhaul to transition to the Julia doc system.

### PR DESCRIPTION
**Pull request by vchuravy**
The plan is to use Documenter.jl for the external documentation.

Right now I am focusing on making REPL documentation available. This also transitions us away from evaling functions at runtime to defining the entry points statically and getting the `MX_handle` dynamically. The overhead should be ok and this gives us better precompilation. Also when `libmxnet.so` changes we automatically recompile `MXNet.jl`

### CODE REVIEW

File: src/ndarray.jl [hunk 1]
```diff
@@ -988,116 +988,143 @@ object (:class:`NDArray`) is returned. Otherwise, a tuple containing all the out
...
+  @mxcall(:MXFuncGetInfo,
+         (MX_handle, Ref{char_p}, Ref{char_p}, Ref{MX_uint}, Ref{char_pp},
+          Ref{char_pp}, Ref{char_pp}, Ref{char_p}),
+          handle, ref_name, ref_desc, ref_narg, ref_arg_names,
+          ref_arg_types, ref_arg_descs, ref_ret_type)
+
+  name = Symbol(unsafe_wrap(String, ref_name[]))
+
+  desc = unsafe_wrap(String, ref_desc[]) * "\n\n"
+  desc *= _format_docstring(Int(ref_narg[]), ref_arg_names, ref_arg_types, ref_arg_descs)
```
Comment (pluskid) [inline src/ndarray.jl]: You might also need to modify `_format_docstring` a bit, as it is written for the ReST syntax.

**Comment by pluskid** [comment]
LGTM

**Comment by vchuravy** [comment]
Yeah this is still quite work in progress and to make it work nicely it relies on a feature that is missing in Documenter. I could splice out the changes to load things with the macro if you want to have them right now.

**Comment by vchuravy** [comment]
But then we couldn't  generate documentation for a while...

**Comment by pluskid** [comment]
@vchuravy What feature of Documenter we are missing? Are they planning to add support for that soon?

**Comment by vchuravy** [comment]
Mostly https://github.com/JuliaDocs/Documenter.jl/issues/72 so that the api documentation will be generated automatically.

**Comment by pluskid** [comment]
@vchuravy it seems the issue in Documenter.jl has been closed. Is the required feature implemented?

**Comment by vchuravy** [comment]
Yes I will get back to this today or tomorrow :)

**Comment by pluskid** [comment]
Is this supposed to be merged or are you still working on this?

**Comment by vchuravy** [comment]
Ok so this is mergable now.

The only thing missing is the setup of the gh-pages branch and travis. Everything else should be in place.

**Comment by pluskid** [comment]
Thanks for the efforts! I tried to configure the keys, and it seem to be working now. The link tot he doc is updated in the README. Many of the docs still have formatting issue, which will need to be manually fixed by us gradually.

**Comment by vchuravy** [comment]
Thanks. yeah I tried to minimise formatting issues, but there is still a lot to do.

**Pull request merged**